nav2_costmap_2d/plugins/costmap_filters/keepout_filter.cpp

```diff
@@ -0,0 +1,256 @@
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Alexey Merzlyakov
+ *********************************************************************/
+
+#include "nav2_costmap_2d/costmap_filters/keepout_filter.hpp"
+
+namespace nav2_costmap_2d
+{
+
+KeepoutFilter::KeepoutFilter()
+: filter_info_sub_(nullptr), mask_sub_(nullptr), mask_costmap_(nullptr)
+{
+}
+
+void KeepoutFilter::initializeFilter(
+  const std::string & filter_info_topic)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  filter_info_topic_ = filter_info_topic;
+  // Setting new costmap filter info subscriber
+  RCLCPP_INFO(
+    node->get_logger(),
+    "Subscribing to \"%s\" topic for filter info...",
+    filter_info_topic.c_str());
+  filter_info_sub_ = node->create_subscription<nav2_msgs::msg::CostmapFilterInfo>(
+    filter_info_topic, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&KeepoutFilter::filterInfoCallback, this, std::placeholders::_1));
+}
+
+void KeepoutFilter::filterInfoCallback(
+  const nav2_msgs::msg::CostmapFilterInfo::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  // Resetting previous subscriber each time when new costmap filter information arrives
+  if (mask_sub_) {
+    RCLCPP_WARN(
+      node->get_logger(),
+      "New costmap filter info arrived from %s topic. Updating old filter info.",
+      filter_info_topic_.c_str());
+    mask_sub_.reset();
+  }
+
+  // Checking that base and multiplier are set to their default values
+  if (msg->base != BASE_DEFAULT or msg->multiplier != MULTIPLIER_DEFAULT) {
+    RCLCPP_ERROR(
+      node->get_logger(),
+      "base and multiplier in ConstmapFilterInfo should be set to their default values (%d and %d)",
```

Comment:
Add "for proper use of the keep-out filter."

Make explicit that this is unique to this plugin

Reply:
Added in https://github.com/ros-planning/navigation2/commit/ab3f5fe2e5f47906832bfee91c2b6ff8e280eb8d#diff-f67a16397bb5979b7cbf4729ea53a451R92